### PR TITLE
add htproxy

### PR DESCRIPTION
**Pull request by maditya**
@andres-erbsen @daniel-ziegler please review!
As discussed with @adon-at-work , added /lookup and /update endpoints which talk json.
Code is very similar to what we have for hkp.

### CODE REVIEW

**Comment by adon-at-work** [comment]
LGTM. :+1:

File: htproxy/htproxy.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+			return
+		}
+	} else if path == "/update" {
+		pf, err = h.doUpdate(r.Body, ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		}
+	}
+
+	marshaler := jsonpb.Marshaler{}
```
Comment (adon-at-work) [inline htproxy/htproxy.go]: We can leave it as is for now. I see potentials that this HTTP proxy can be less "dump" by making the lookupProof data more friendly and favorable to the JS client.

https://github.com/yahoo/end-to-end/blob/feature/lookup-client-v2/src/javascript/crypto/e2e/coname/client.js#L188-L230 shows a list of things that can be moved from the client to this particular position at the server.

File: htproxy/htproxy.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, err
+	}
+	return pf, nil
+
+}
+
+func (h *HTProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	path := r.URL.Path
+
+	//TODO: auth checks would go here
```
Comment (adon-at-work) [inline htproxy/htproxy.go]: My two cents. Can we create an auth stub in a separate file, e.g., auth.go? Therefore, i'm anticipating `Auth.verify(r)` which always return true. That would help others to patch the auth.go file amid easily merging latest changes to htproxy.go

**Comment by adon-at-work** [comment]
In general, LGTM. :+1: 

- I like the idea of having a http interface that is dedicated for serving the JS client. So, we can later improve by incorporating some optimizations/improvements that make the JS client thinner/happier. :)

File: htproxy/htproxy.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	if path == "/lookup" {
+		pf, err = h.doLookup(r.Body, ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	} else if path == "/update" {
+		pf, err = h.doUpdate(r.Body, ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (adon-at-work) [inline htproxy/htproxy.go]: looks like you missed a `return` here

File: htproxy/htproxy.go (hunk 2 above)
Comment (maditya) [inline htproxy/htproxy.go]: sure, done with https://github.com/yahoo/coname/commit/4825be0c1820efe2ad397d580aa41055b0fca0d1

File: htproxy/htproxy.go (hunk 3 above)
Comment (maditya) [inline htproxy/htproxy.go]: fixed!

**Comment by andres-erbsen** [comment]
I don't think I am up to date with the authentication story; when I first read this I was still under the impression that initial registration is authenticated by DKIM and further updates are authenticated inside the keyserver. Is there any place where I could read up on this?

Is saying `htproxy` instead of `httpproxy` or `httpfront` meant to signify anything?

Otherwise, I think the HKPProxy-like architecture makes a lot of sense and so does the code.

**Comment by adon-at-work** [comment]
This PR is largely meant for exposing an HTTP interface for the extension to talk to. `httpfront` sounds better to me, but let's not care about how it is named. @maditya mentioned he preferred to keep the name as short as possible. 

The HTTP agent is now sending authentication Cookie to the htproxy, which is hosted at `somewhere.yahoo.com`. The htproxy can thus authenticate the user with the cookie presented. Given such cookie is tied to a specific user account name and a single email address (one-to-one mapping assumed), it's already sufficient as an auth. means to let users update their keys.

In addition to the auth above, the server can uses the DKIM headers included in the update message for further authentication (auth additional email addresses) as/if needed. However, we'd avoid using DKIM for now for simplicity (automatically capturing DKIM headers from webmail is difficult from the extension perspective. in fact, the headers by default are always hidden from the users/extension).

**Comment by maditya** [comment]
renamed to httpfront. user auth story is something we're discussing, lets take it offline since its kind of unrelated to this PR.

**Pull request merged**